Add vscode tasks to run tests in current file/project

RikkiGibson opened the pull request:
This PR adds a couple of vscode tasks to improve developer productivity.

- run tests in current file
- run tests in current project

(running all tests is still best handled by Test.cmd)

The task will also write the test results to an html file in TestResults. This should be useful for keeping track of which tests were failing or sifting through test runs with many failures.

I didn't make the task try to search for .vbproj files because VB doesn't work that well in vscode. You are better off going to full VS if you need to work in VB.

Here's a demo:
![demo-test-task-3](https://user-images.githubusercontent.com/5833655/79518143-2b6d5a00-8005-11ea-88e9-f43189ea13b5.gif)

jmarolf commented:
What is the recommendation for tasks like these if we want them to run on mac and linux? Is there a way to call the bash versions of these in those cases?

RikkiGibson commented:
vscode does have an answer for this: https://code.visualstudio.com/Docs/editor/tasks#_operating-system-specific-properties.

however that seems to imply writing the script again in bash. is that the right thing to do? any chance of using cross platform powershell for this?

jaredpar commented on scripts/vscode-run-tests.ps1:
```diff
@@ -0,0 +1,56 @@
+Set-Location $fileInfo.Directory
+
+try
+{
```
Nit: Powershell coding style puts braces on same line

jaredpar commented on scripts/vscode-run-tests.ps1:
```diff
@@ -0,0 +1,56 @@
+        if ($files)
+        {
+            Pop-Location
+            $dotnetPath = Resolve-Path "$PSScriptRoot/../.dotnet/dotnet.exe" -Relative
```
Did u consider importing the build-utils script which has our definitive way of grabbing the dotnet binary?

333fred replied:
Please do this so I can use this on Linux.

333fred commented:
> however that seems to imply writing the script again in bash.

pwsh is available for linux, and is how I generate compiler code. Can you add using that for Linux and I'll give it a try?

RikkiGibson commented:
Thanks for the feedback. I'll pull it down on my Mac and make sure it works nicely on there.

This test failure is a new one to me: `Microsoft.CodeAnalysis.CSharp.UnitTests.Symbols.GenericConstraintTests.DefaultT(): System.BadImageFormatException : Operation did not complete successfully because the file contains a virus or potentially unwanted software. (0x800700E1)`

RikkiGibson commented:
I opted for a solution that uses a local 'dotnet tool' install of PowerShell. It requires running 'dotnet tool restore' in the repo root before the commands will work. See what you think.

333fred commented:
> I opted for a solution that uses a local 'dotnet tool' install of PowerShell. It requires running dotnet tool restore in the repo root before the commands will work. See what you think.

Personally I already have pwsh installed, but that's really w/e, I only have it installed for this repo.

RikkiGibson commented:
solved 2 lingering issues:

1. tasks no longer use the globally installed dotnet cli, instead use the dotnet cli installed as part of restore. this works on windows and mac and presumably on linux.
2. dotnet tools are restored upon `.\Restore.cmd` or `./restore.sh`. so if you just clone the repo, run a restore, pop open vscode, you should be able to start running tasks to your heart's content.

I think this is in a pretty good place now. The test running scripts in particular I am finding invaluable when working in vscode. I also extracted out the bit that finds the project file and modified the "build directory" task to just build the "current project" based on what file you have open. /cc @jaredpar

333fred commented on dotnet-tools.json:
```diff
@@ -2,10 +2,16 @@
       ]
+    },
+    "powershell": {
+      "version": "6.2.4",
```
7.0 has been out for a month, we should probably use that.
